File: java/TrappingRainWater.java
public class TrappingRainWater {
    public int trap(int[] bars) {
        if (bars == null) {
            return 0;
        }
        int n = bars.length;
        if (n == 0) {
            return 0;
        }
        int[] leftHighest = new int[n];
        int[] rightHighest = new int[n];
        leftHighest[0] = bars[0];
        for (int i = 1; i < n; ++i) {
            leftHighest[i] = Math.max(leftHighest[i - 1], bars[i]);
        }
        rightHighest[n - 1] = bars[n - 1];
        for (int i = n - 2; i >= 0; --i) {
            rightHighest[i] = Math.max(rightHighest[i + 1], bars[i]);
        }
        int sum = 0;
        for (int i = 1; i < n - 1; ++i) {
            sum += Math.min(leftHighest[i], rightHighest[i]) - bars[i];
        }
        return sum;
    }
}

File: python/trapping_rain_water.py
#!/usr/bin/env python3

class Solution:
    def trap(self, A):
        if A is None:
            return 0
        n = len(A)
        if n == 0:
            return 0
        left_high = [0] * n
        left_high[0] = A[0]
        for i in range(1, n):
            left_high[i] = max(left_high[i - 1], A[i])
        right_high = [0] * n
        right_high[n - 1] = A[n - 1]
        for i in reversed(range(0, n - 1)):
            right_high[i] = max(right_high[i + 1], A[i])
        res = 0
        for i in range(1, n - 1):
            res = res + max(0, min(left_high[i - 1], right_high[i + 1]) - A[i])
        return res

def main():
    solver = Solution()
    A = [0,1,0,2,1,0,1,3,2,1,2,1]
    print(solver.trap(A))
    A = [0]
    print(solver.trap(A))
    A = [0, 1]
    print(solver.trap(A))
    A = [2, 1, 2]
    print(solver.trap(A))
    A = None
    print(solver.trap(A))
    A = []
    print(solver.trap(A))

if __name__ == '__main__':
    main()
